File: app/src/main/java/com/security/passwordmanager/EmailPasswordActivity.java
package com.security.passwordmanager;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.util.PatternsCompat;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.FirebaseDatabase;
import com.security.passwordmanager.ui.main.PasswordListActivity;

import java.util.Objects;

public class EmailPasswordActivity extends AppCompatActivity implements View.OnClickListener {

    private static final String TAG = "EmailPassword";

    private ProgressBar progressBar;
    private Support mSupport;
    private EditText mEmailField, mPasswordField;
    private Button signIn;
    private CheckBox isPasswordRemember;

    private TextView label, subtitle;

    private FirebaseAuth mAuth;

    public static Intent getIntent(Context context) {
        Intent intent = new Intent(context, EmailPasswordActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
        return intent;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mAuth = FirebaseAuth.getInstance();
        mSupport = Support.getInstance(this);

        label = findViewById(R.id.text_view_main_label);
        subtitle = findViewById(R.id.text_view_main_subtitle);

        mEmailField = findViewById(R.id.username);
        mPasswordField = findViewById(R.id.password);
        signIn = findViewById(R.id.signIn);
        Button signUp = findViewById(R.id.signUp);

        isPasswordRemember = findViewById(R.id.remember_password);
        isPasswordRemember.setChecked(mSupport.isPasswordRemembered());

        if (isPasswordRemember.isChecked())
            startActivity(PasswordListActivity.getIntent(this));

        isPasswordRemember.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked && (TextUtils.isEmpty(mPasswordField.getText()) ||
                            TextUtils.isEmpty(mEmailField.getText())))
                isPasswordRemember.setChecked(false);

            mSupport.setPasswordIsRemembered(isChecked);
        });

        progressBar = findViewById(R.id.loading);

        signIn.setOnClickListener(this);
        signUp.setOnClickListener(this);

        addTextChangeListener(mEmailField, true);
        addTextChangeListener(mPasswordField, false);
    }

    @Override
    protected void onResume() {
        super.onResume();
        progressBar.setVisibility(View.GONE);
        updateUI();
    }

    @Override
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        MenuItem search = menu.findItem(R.id.menu_item_search);
        search.setVisible(false);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.menu_item_settings)
            startActivity(SettingsActivity.newIntent(this, false));
        return super.onOptionsItemSelected(item);
    }



    private void addTextChangeListener(EditText view, boolean email) {
        view.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (email)
                    validateForm(s.toString(), null);
                else
                    validateForm(null, s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });
    }



    private void updateUI() {
        mSupport.updateThemeInScreen(getWindow(), Objects.requireNonNull(getSupportActionBar()));

        label.setTextColor(mSupport.getFontColor());
        subtitle.setTextColor(mSupport.getFontColor());

        signIn.setBackgroundResource(mSupport.getButtonRes());

        mEmailField.setTextColor(mSupport.getFontColor());
        mPasswordField.setTextColor(mSupport.getFontColor());

        mEmailField.setHintTextColor(mSupport.getDarkerGrayColor());
        mPasswordField.setHintTextColor(mSupport.getDarkerGrayColor());

        mEmailField.setBackgroundTintList(ColorStateList.valueOf(mSupport.getFontColor()));
        mPasswordField.setBackgroundTintList(ColorStateList.valueOf(mSupport.getFontColor()));

        isPasswordRemember.setTextColor(mSupport.getFontColor());
        isPasswordRemember.setButtonTintList(ColorStateList.valueOf(mSupport.getHeaderColor()));
    }


    private void registerUser() {
        String email = mEmailField.getText().toString().trim();
        String password = mPasswordField.getText().toString().trim();
        if (!validateForm(email, password)) {
            progressBar.setVisibility(View.GONE);
            return;
        }

        mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        mAuth = FirebaseAuth.getInstance();
                        FirebaseDatabase.getInstance().getReference("Users")
                                .child(Objects.requireNonNull(mAuth.getCurrentUser()).getUid())
                                .setValue(email)
                                .addOnCompleteListener(task1 -> {
                                    if (task1.isSuccessful()) {
                                        Toast.makeText(
                                                this,
                                                R.string.register_successful,
                                                Toast.LENGTH_LONG).show();

                                        startActivity(PasswordListActivity.getIntent(this));
                                    }
                                    else
                                        Toast.makeText(
                                                this,
                                                R.string.register_failed,
                                                Toast.LENGTH_LONG).show();

                                    progressBar.setVisibility(View.GONE);
                                });
                    }
                    else {
                        Toast.makeText(
                                this,
                                R.string.register_failed,
                                Toast.LENGTH_LONG).show();

                        progressBar.setVisibility(View.GONE);
                    }
                });
    }



    private void loginUser() {
        String email = mEmailField.getText().toString().trim();
        String password = mPasswordField.getText().toString().trim();

        if (!validateForm(email, password)) {
            progressBar.setVisibility(View.GONE);
            return;
        }

        mAuth.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful())
                        startActivity(PasswordListActivity.getIntent(this));
                    else
                        Toast.makeText(
                                this,
                                R.string.login_failed,
                                Toast.LENGTH_LONG).show();
        });
    }


    private boolean validateForm(@Nullable String email, @Nullable String password) {

        if (email != null) {
            if (TextUtils.isEmpty(email)) {
                mEmailField.setError(getString(R.string.required));
                mEmailField.requestFocus();
                return false;
            } else if (!PatternsCompat.EMAIL_ADDRESS.matcher(email).matches()) {
                mEmailField.setError(getString(R.string.valid_email));
                mEmailField.requestFocus();
                return false;
            } else
                mEmailField.setError(null);
        }


        if (password != null) {
            if (TextUtils.isEmpty(password)) {
                mPasswordField.setError(getString(R.string.required));
                mPasswordField.requestFocus();
                return false;
            }
            else if (password.length() < 6) {
                mPasswordField.setError(getString(R.string.min_password_length));
                mPasswordField.requestFocus();
                return false;
            }
            else
                mPasswordField.setError(null);
        }

        return true;
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public void onClick(View v) {
        progressBar.setVisibility(View.VISIBLE);

        switch (v.getId()) {
            case R.id.signUp:
                registerUser();
                break;
            case R.id.signIn:
//                startActivity(PasswordListActivity.getIntent(this));
                loginUser();
                break;
        }
    }
}

File: app/src/main/java/com/security/passwordmanager/EmailPasswordActivity.kt
package com.security.passwordmanager

import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.os.Bundle
import android.text.InputType
import android.text.TextUtils
import android.view.View
import android.widget.CompoundButton
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.util.PatternsCompat
import androidx.core.widget.doOnTextChanged
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.security.passwordmanager.databinding.ActivityMainBinding
import com.security.passwordmanager.settings.SettingsViewModel
import com.security.passwordmanager.ui.main.PasswordListActivity

class EmailPasswordActivity : AppCompatActivity(), View.OnClickListener {

    private fun TextView.isEmpty() =
        TextUtils.isEmpty(text)

    companion object {
//        private const val TAG = "EmailPassword"

        fun getIntent(context: Context) : Intent {
            val intent = Intent(context, EmailPasswordActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
            return intent
        }
    }

    private lateinit var settings: SettingsViewModel

    private lateinit var binding: ActivityMainBinding
    private lateinit var mAuth: FirebaseAuth

    private var isPasswordVisible = false

    override fun onCreate(savedInstanceState: Bundle?) {
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
        binding = ActivityMainBinding.inflate(layoutInflater)

        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        mAuth = FirebaseAuth.getInstance()
        settings = SettingsViewModel.getInstance(this)

        binding.rememberPassword.isChecked = settings.isPasswordRemembered

        if (binding.rememberPassword.isChecked)
            startActivity(PasswordListActivity.getIntent(this))

        binding.rememberPassword.setOnCheckedChangeListener {_: CompoundButton?, isChecked: Boolean ->
            if (isChecked &&
                (binding.password.isEmpty() || binding.login.isEmpty()))
                    binding.rememberPassword.isChecked = false

            settings.isPasswordRemembered = isChecked
        }

        binding.signIn.setOnClickListener(this)
        binding.signUp.setOnClickListener(this)

        binding.login.addTextChangedListener(true)
        binding.password.addTextChangedListener(false)

        binding.passwordVisibility.setOnClickListener {
            updatePasswordView()
        }
    }


    override fun onResume() {
        super.onResume()
        binding.loading.hide()
        updateUI()
    }


    private fun EditText.addTextChangedListener(email : Boolean) {
        doOnTextChanged { text, _, _, _ ->
            if (email)
                validateForm(text.toString(), null)
            else
                validateForm(null, text.toString())
        }
    }


    private fun updateUI() {
        settings.updateThemeInScreen(window, supportActionBar)

        settings.fontColor.let {
            binding.mainLabel.setTextColor(it)
            binding.mainSubtitle.setTextColor(it)
            binding.login.setTextColor(it)
            binding.password.setTextColor(it)
            binding.rememberPassword.setTextColor(it)
            binding.login.backgroundTintList = ColorStateList.valueOf(it)
            binding.password.backgroundTintList = ColorStateList.valueOf(it)

            binding.passwordVisibility.imageTintList = ColorStateList.valueOf(it)
        }

        binding.signIn.setBackgroundResource(settings.buttonRes)
        binding.passwordVisibility.setBackgroundColor(settings.backgroundColor)
        binding.rememberPassword.buttonTintList = ColorStateList.valueOf(settings.headerColor)
    }

    private fun updatePasswordView() = binding.run {
        isPasswordVisible = !isPasswordVisible

        if (isPasswordVisible) {
            password.inputType = InputType.TYPE_CLASS_TEXT
            passwordVisibility.setImageResource(R.drawable.visibility_off)
            passwordVisibility.contentDescription = getString(R.string.hide_password)
        }
        else {
            password.inputType = 129
            passwordVisibility.setImageResource(R.drawable.visibility_on)
            passwordVisibility.contentDescription = getString(R.string.show_password)
        }
    }




    private fun registerUser() {
        val email = binding.login.text.toString().trim()
        val password = binding.password.text.toString().trim()

        if (!validateForm(email, password)) {
            binding.loading.hide()
            return
        }

        mAuth.createUserWithEmailAndPassword(email, password)
            .addOnCompleteListener { task: Task<AuthResult?> ->
                if (task.isSuccessful) {
                    mAuth = FirebaseAuth.getInstance()

                    mAuth.currentUser?.uid?.let {
                        FirebaseDatabase.getInstance().getReference("Users")
                            .child(it)
                            .setValue(email)
                            .addOnCompleteListener { task1: Task<Void?> ->
                                if (task1.isSuccessful) {
                                    Toast.makeText(
                                        this,
                                        R.string.register_successful,
                                        Toast.LENGTH_LONG
                                    ).show()

                                    startActivity(PasswordListActivity.getIntent(this))

                                } else Toast.makeText(
                                    this,
                                    R.string.register_failed,
                                    Toast.LENGTH_LONG
                                ).show()
                                binding.loading.hide()
                            }
                    }
                } else {
                    Toast.makeText(
                        this,
                        R.string.register_failed,
                        Toast.LENGTH_LONG
                    ).show()
                    binding.loading.hide()
                }
            }
    }


    private fun loginUser() {
        val email = binding.login.text.toString().trim()
        val password = binding.password.text.toString().trim()

        if (!validateForm(email, password)) {
            binding.loading.hide()
            return
        }

        mAuth.signInWithEmailAndPassword(email, password)
            .addOnCompleteListener { task: Task<AuthResult?> ->
                if (task.isSuccessful)
                    startActivity(PasswordListActivity.getIntent(this))
                else Toast.makeText(
                    this,
                    R.string.login_failed,
                    Toast.LENGTH_LONG
                ).show()
            }
    }

    private fun validateForm(email: String?, password: String?): Boolean {

        if (email == null && password == null) return false

        if (email != null) {
            when {
                email.isEmpty() -> {
                    binding.login.error = getString(R.string.required)
                    binding.login.requestFocus()
                    return false
                }
                !PatternsCompat.EMAIL_ADDRESS.matcher(email).matches() -> {
                    binding.login.error = getString(R.string.valid_email)
                    binding.login.requestFocus()
                    return false
                }
                else -> binding.login.error = null
            }
        }

        if (password != null) when {
            password.isEmpty() -> {
                binding.password.error = getString(R.string.required)
                binding.password.requestFocus()
                return false
            }
            password.length < 6 -> {
                binding.password.error = getString(R.string.min_password_length)
                binding.password.requestFocus()
                return false
            }
            else -> binding.password.error = null
        }
        return true
    }


    override fun onClick(v: View?) {
        binding.loading.show()

        when(v?.id) {
            R.id.sign_up -> registerUser()
            R.id.sign_in -> loginUser()
        }
    }
}